### ISSUE TITLE
Concurrent transactions

### ISSUE THREAD

**Issue by Kixiron** [vmware/differential-datalog#823]
My application is heavily concurrent, and at data-introduction time there's a number of threads attempting to dump data into ddlog. With the current transaction system I have transactions protected by a mutex and I only lock the mutex for the `.apply_valupdates()` portion, but still with those measures in place larger sets of data being introduced will effectively make the system progress in serial since other threads are waiting to be able to introduce their data. A transaction system that allows multiple concurrent transactions would solve this, allowing for a higher level of parallelism

**Comment by ryzhyk** [comment]
There are several issues intertwined here.  Let me try to break things up.  First, there is the problem of scaling DDlog with the number of threads (see, e.g., #820). Ideally, given a large enough transaction, DDlog should be able to process it faster on multiple cores using multiple DD workers.  Note that this does not require support for concurrent transactions, rather this is a way to achieve parallelism inside a transaction.

Internal parallelism aside, I agree that a more concurrent API would make sense. I think such an API can be either built on top of what we've got today or added using relatively minor modifications to DDlog internals, but one has to be careful in defining the semantics of concurrent transactions, as it is not completely intuitive and if not approached carefully may lead to surprising behavior. The first question is whether you want transactions to be truly concurrent, i.e., changes made by concurrent transactions logically belong to the same timestamp. This design probably won't work well, as (1) no transaction can commit until all concurrent transactions have committed, as there is no way to isolate outputs produced by different transactions, (2) when program inputs are declared as `relations` (i.e., sets) as opposed to `streams` of `multisets`,  transactions can conflict, and I really don't want to worry about conflict resolution, rollbacks due to conflicts and all that.

Another option is to run logically serialized transactions in parallel.  When a transaction starts, it obtains a monotonically increasing serial number, which corresponds to a DD epoch. Any number of transactions can push data in parallel.  The data gets cached either inside DDlog or inside DD (the latter will require changing DDlog internals slightly).  When a transaction calls commit, it blocks until this ***and all transactions with smaller timestamps*** commit.   Is this what had in mind?

**Comment by Kixiron** [comment]
Ideally the first would be implemented since each of my transactions are logically separate in all aspects (that is, data does not cross between them), but either could work. My primary purpose is speed since as of now serial transactions are massively crippling execution speed. Internal scaling would also be a huge boon, I'd be happy to help with getting that implemented if you gave some pointers

**Comment by ryzhyk** [comment]
If your transactions are truly isolated, would it make sense to split your DDlog program into multiple independent programs?

**Comment by Kixiron** [comment]
They're isolated at the moment, but implementing cross-file analysis is planned so that'd only be a temporary solution

**Comment by ryzhyk** [comment]
> They're isolated at the moment, but implementing cross-file analysis is planned so that'd only be a temporary solution

But won't you then run into the problems with truly concurrent transactions that I outlined above?

**Comment by Kixiron** [comment]
Good point, sequential transactions sounds good then, maybe offer a non-blocking option that returns a handle that can later be blocked on?

**Comment by Kixiron** [comment]
I did some benchmarking and you're completely correct, threads are very detrimental here
```
W1 dataflow operator execution: 189.654s
W1 Non-nested operators execution: 186.715s
W1 Guarded seconds: 59.652s

W2 dataflow operator execution: 328.240s
W2 Non-nested operators execution: 317.322s
W2 Guarded seconds: 109.676s
```

**Comment by namibj** [comment]
> Another option is to run logically serialized transactions in parallel. When a transaction starts, it obtains a monotonically increasing serial number, which corresponds to a DD epoch. Any number of transactions can push data in parallel. The data gets cached either inside DDlog or inside DD (the latter will require changing DDlog internals slightly). When a transaction calls commit, it blocks until this _**and all transactions with smaller timestamps**_ commit. Is this what had in mind?
This sounds quite promising.

I'm thinking of just duplicating Timely [`Handle`](https://docs.rs/timely/0.11.1/timely/dataflow/operators/input/struct.Handle.html)'s for the individual transactions, and dropping them (later) on `commit`.
[This](https://docs.rs/differential-dataflow/0.11.0/differential_dataflow/input/struct.InputSession.html#method.from) method re-creates a DDflow `InputSession` from the Handles.
Just one "little" issue. This is ordered input, which doesn't allow nice, concurrent feeding.

Timely's [`UnorderedInput`](https://docs.rs/timely/0.11.1/timely/dataflow/operators/unordered_input/trait.UnorderedInput.html) seems to need a bit more manual coding to properly feed DDflow's own timestamps, but (to my understanding) DDlog doesn't really use DDflow's timestamp-combining abilities.
But it produces a normal `Stream` that works just as usual.
The intra-transaction ordering, if needed, could probably be provided by another `Timestamp` layer, if necessary.

The Futures would just poll on (a) probe(s) of the output(s) to signal that the specific transaction's output is only complete when the timestamp shows up in the probe(s).
